lamindb/models/history.py

```diff
@@ -0,0 +1,89 @@
+from django.db import models
+
+
+class HistoryTableState(models.Model):
+    """A list of tables for which we're tracking history.
+
+    This table serves two purposes: it allows us to store the
+    name of the table for a given history record with less state.
+
+    NOTE: This table only makes sense for a single LaminDB instance,
+    since the ID of a given table may be different on different instances.
+    You should join this table into History to determine which table
+    a given history record applies to, rather than using the IDs directly.
+    """
+
+    # Using a smallserial here because it's incredibly unlikely that
+    # we'll have more than 32k tables
+    id = models.SmallAutoField(primary_key=True)
+    table_name = models.CharField(max_length=255)
+    backfilled = models.BooleanField()
+
+
+class HistoryMigrationState(models.Model):
+    """A summary of the state of Django's migrations when a history record was recorded.
+
+    When a history record is recorded, we need to record the state of the data migrations
+    for each Django app so that we know if we need to migrate the record from the version
+    it had when it was written to the version that exists in the target during replay.
+    """
+
+    # Using a smallserial here because it's incredibly unlikely that
+    # we'll have more than 32k schema migrations.
+    id = models.SmallAutoField(primary_key=True)
+    migration_history_id = models.JSONField()
+
+
+class History(models.Model):
+    """Stores the edit history for all LaminDB tables."""
+
+    seqno = models.AutoField(primary_key=True)
+    history_migration_state = models.ForeignKey(
+        HistoryMigrationState, on_delete=models.PROTECT
+    )
+    table = models.ForeignKey(HistoryTableState, on_delete=models.PROTECT)
+    id = models.UUIDField()
+    # Many-to-many tables don't have row UIDs, so this needs to be nullable.
+    record_uid = models.JSONField(null=True)
+    record_data = models.JSONField(null=True)
+    event_type = models.PositiveSmallIntegerField()
+    created_at = models.DateTimeField()
```

Comment:
We should add a `space_uid` that's based on the `record.space.uid` of the record whose history is logged. This makes a great chunk of access management for the history table easier (RLS is easier). It should be trivial because you have logic for this already above.

We should add a `created_by_uid` string field that's automatically populated based on the database connection that codes the user identity. Lamin Team can handle this.

Reply:
Please add a `branch_code = models.SmallIntegerField()`.

Reply:
Please add a `run_uid = CharField(max_length=20)`.